use `on_unimplemented` for better error messages

sahandevs opened the pull request:
closes #2 

I don't have an easy way to write a UI test for this. Any ideas?

![image](https://user-images.githubusercontent.com/1113944/186654405-c8e71c85-4566-4728-83e6-d55e9f387273.png)

maminrayej commented on src/lib.rs:
```diff
@@ -399,7 +395,23 @@ pub fn builder(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
                         }
                     }
 
+                    let mut error_message = vec![];
+                    // this feature uses `#[rustc_on_unimplemented]` which is only available
+                    // in a nightly compiler.
+                    if cfg!(feature = "better_error_message") {
+                        let message = format!("missing `{}`", &field_name);
+                        let label =
+                            format!("provide `{}` before calling `.build()`", &field_name);
+                        error_message.push(quote! {
+                            #[rustc_on_unimplemented(
+                                message=#message,
+                                label=#label,
+                            )]
+                        });
+                    }
+
                     constraint_traits.push(quote! {
+                        #(#error_message)*
                         trait #trait_name {}
```
Shouldn't we have a single message per `constraint_trait`? Why `error_message` is a `vec`?

maminrayej commented on Cargo.toml:
```diff
@@ -9,6 +9,9 @@ license = "MIT"
 keywords = ["builder", "builder_pattern"]
 
+[features]
+better_error_message = []
```
May I suggest a more concise name?
How about `better_error`? It kinda rhythms :)

maminrayej commented:
I didn't understand your problem with writing a UI test unfortunately. Can you elaborate?

sahandevs commented:
Ok! I've sold my soul to the devil to add some tests. :))
also, I've addressed the issues that you mentioned. would you please review it again? @maminrayej

maminrayej commented:
Wow you really went above and beyond to make it work huh? :))
All is well and I'll merge this as is, but I want to discuss the matter here a bit for future reference.

It's my understanding that not being able to tell `cargo` which feature should be enabled for which test is our main problem right?
Also we can't run some tests using stable Rust and some other using nightly(using one `cargo` command).

sahandevs commented:
Exactly, we can't specify crate features and compiler targets using the `try_build` crate.

maminrayej commented:
@sahandevs Could you look at this [branch](https://github.com/maminrayej/tidy-builder/tree/split-tests) as an alternative?
It uses the names of the test files to distinguish stable tests from nightly tests, and uses a `Makefile` to make running the tests easier. It's still a hacky way to do things, but it extracts the logic of splitting tests from the tests themselves (the `ui.rs` file does not need to spawn a new `cargo test`) and delegates it to a `Makefile`.
